pkg/target/manager.go

```diff
@@ -0,0 +1,527 @@
+
+	gardencore "github.com/gardener/gardener/pkg/apis/core"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	ErrNoGardenTargeted  = errors.New("no garden cluster targeted")
+	ErrNoProjectTargeted = errors.New("no project targeted")
+	ErrNoSeedTargeted    = errors.New("no seed cluster targeted")
+	ErrNoShootTargeted   = errors.New("no shoot targeted")
+)
+
+type Manager interface {
+	CurrentTarget() (Target, error)
+
+	TargetGarden(name string) error
+	TargetProject(ctx context.Context, name string) error
+	TargetSeed(ctx context.Context, name string) error
+	TargetShoot(ctx context.Context, name string) error
+
+	GardenClient(t Target) (client.Client, error)
+	ProjectClient(t Target) (client.Client, error)
+	SeedClient(t Target) (client.Client, error)
+	ShootNamespaceClient(t Target) (client.Client, error)
+	ShootClusterClient(t Target) (client.Client, error)
+}
+
+type managerImpl struct {
+	config          *config.Config
+	targetProvider  TargetProvider
+	clientProvider  ClientProvider
+	kubeconfigCache KubeconfigCache
+}
+
+var _ Manager = &managerImpl{}
+
+func NewManager(config *config.Config, targetProvider TargetProvider, clientProvider ClientProvider, kubeconfigCache KubeconfigCache) (Manager, error) {
+	return &managerImpl{
+		config:          config,
+		targetProvider:  targetProvider,
+		clientProvider:  clientProvider,
+		kubeconfigCache: kubeconfigCache,
+	}, nil
+}
+
+func (m *managerImpl) CurrentTarget() (Target, error) {
+	return m.targetProvider.Read()
+}
+
+func (m *managerImpl) TargetGarden(gardenName string) error {
+	for _, g := range m.config.Gardens {
+		if g.Name == gardenName {
+			return m.patchTarget(func(t *targetImpl) error {
+				t.Garden = gardenName
+				t.Project = ""
+				t.Seed = ""
+				t.Shoot = ""
+
+				return nil
+			})
+		}
+	}
+
+	return fmt.Errorf("garden %q is not defined in gardenctl configuration", gardenName)
+}
+
+func (m *managerImpl) TargetProject(ctx context.Context, projectName string) error {
+	return m.patchTarget(func(t *targetImpl) error {
+		if t.Garden == "" {
+			return ErrNoGardenTargeted
+		}
+
+		gardenClient, err := m.clientForGarden(t.Garden)
+		if err != nil {
+			return fmt.Errorf("could not create Kubernetes client for garden cluster: %w", err)
+		}
+
+		// validate that the project exists
+		project, err := m.resolveProjectName(ctx, gardenClient, projectName)
+		if err != nil {
+			return fmt.Errorf("failed to resolve project: %w", err)
+		}
+
+		// validate the project
+		if err := m.validateProject(ctx, project); err != nil {
+			return fmt.Errorf("invalid project: %w", err)
+		}
+
+		t.Seed = ""
+		t.Project = projectName
+		t.Shoot = ""
+
+		return nil
+	})
+}
+
+func (m *managerImpl) resolveProjectName(ctx context.Context, gardenClient client.Client, projectName string) (*gardencorev1beta1.Project, error) {
```

Comment:
shouldn't such a function be named `resolveProjectByName`?

Reply:
Thhe function resolves the name to a project. "resolve project name" makes total sense to me. With a "by" I would expect a function name like "_find_ project _by_ name".